### PR TITLE
feat(natives/five): rope getter and setter natives

### PR DESCRIPTION
**Pull request by Hendreik**
Adds GET_ALL_ROPES, with four getters and a setter for ropes. Class structures remain the same between 1604 and 2189

### CODE REVIEW

File: code/components/extra-natives-five/include/ropeManager.h [hunk 1]
```diff
@@ -0,0 +1,141 @@
...
+
+enum eRopeFlags : uint8_t
+{
+	DrawShadowEnabled = 2,
+	Breakable = 4,
+	RopeUnwindingFront = 8,
+	RopeWinding = 32
+};
+
+class rope // No RTTI for this. Unknown class name.
```
Comment (blattersturm) [inline code/components/extra-natives-five/include/ropeManager.h:68]: Companion says it's `rage::ropeInstance`.

File: code/components/extra-natives-five/include/ropeManager.h [hunk 2]
```diff
@@ -0,0 +1,141 @@
...
+	verletTaskManager* verletTaskManager; // 0x10
+	uint32_t numAllocated; // 0x18
+	uint32_t numAvailable; // 0x1C
+	char pad_20[0x8]; // 0x20
+	LinkedList<rope*> allocated; // 0x28
+	LinkedList<rope*> available; // 0x38
+	char pad_48[0x10]; // 0x48
+	LinkedList<phVerletCloth*> cloths; // 0x58
+
+	rope* GetRopeFromHandle(uint32_t handle);
```
Comment (blattersturm) [inline code/components/extra-natives-five/include/ropeManager.h:136]: `rage::ropeManager::FindRope(int)`?

File: code/components/extra-natives-five/src/RopeNatives.cpp [hunk 3]
```diff
@@ -0,0 +1,146 @@
...
+});
+
+rope* ropeManager::GetRopeFromHandle(uint32_t handle)
+{
+	return ropeManager_getRopeFromHandle(this, handle);
+}
+
+static ropeManager** g_ropeManager;
+
+ropeManager* ropeManager::Get()
```
Comment (blattersturm) [inline code/components/extra-natives-five/src/RopeNatives.cpp:30]: GetInstance makes more sense as a name for singletons. Matches some non-inlined stuff in companion as well.

File: code/components/extra-natives-five/src/RopeNatives.cpp [hunk 4]
```diff
@@ -0,0 +1,146 @@
...
+
+#include "ropeManager.h"
+
+#include <Hooking.h>
+#include <ScriptEngine.h>
+#include <ScriptSerialization.h>
+
+namespace rage
+{
+static ropeDataManager* g_ropeDataManager;
```
Comment (blattersturm) [inline code/components/extra-natives-five/src/RopeNatives.cpp:11]: Technically, rage::ropeDataManager::ms_Instance.

File: code/components/extra-natives-five/src/RopeNatives.cpp [hunk 5]
```diff
@@ -0,0 +1,146 @@
...
+
+ropeDataManager* ropeDataManager::Get()
```
Comment (blattersturm) [inline code/components/extra-natives-five/src/RopeNatives.cpp:13]: GetInstance here as well.

File: code/components/extra-natives-five/include/ropeManager.h [hunk 6]
```diff
@@ -0,0 +1,141 @@
...
+class grcTexture;
+
+class ropeData
+{
+public:
+	virtual ~ropeData() = 0;
+
+	virtual const char* GetName() = 0;
+
+	virtual parStructure* GetParStructure() = 0;
```
Comment (blattersturm) [inline code/components/extra-natives-five/include/ropeManager.h:20]: `parser_GetStructure`

**Review by blattersturm — COMMENTED** [review]
Naming comments.

**Comment by Hendreik** [comment]
Thanks, useful information to know

**Pull request merged**